apps/emqx_gateway_nats/src/emqx_nats_channel.erl

```diff
@@ -0,0 +1,1213 @@
+
+%% @doc Init protocol
+init(
+    ConnInfo = #{
+        peername := {PeerHost, _} = PeerName,
+        sockname := {_, SockPort}
+    },
+    Option
+) ->
+    Peercert = maps:get(peercert, ConnInfo, undefined),
+    Mountpoint = maps:get(mountpoint, Option, undefined),
+    ListenerId =
+        case maps:get(listener, Option, undefined) of
+            undefined -> undefined;
+            {GwName, Type, LisName} -> emqx_gateway_utils:listener_id(GwName, Type, LisName)
+        end,
+    EnableAuthn = maps:get(enable_authn, Option, true),
+    ClientInfo = setting_peercert_infos(
+        Peercert,
+        #{
+            zone => default,
+            listener => ListenerId,
+            protocol => nats,
+            peerhost => PeerHost,
+            peername => PeerName,
+            sockport => SockPort,
+            clientid => emqx_utils:rand_id(?RAND_CLIENTID_BYETS),
+            username => undefined,
+            is_bridge => false,
+            is_superuser => false,
+            enable_authn => EnableAuthn,
+            mountpoint => Mountpoint
+        }
+    ),
+
+    Ctx = maps:get(ctx, Option),
+    Override = maps:merge(
+        ?DEFAULT_OVERRIDE,
+        maps:get(clientinfo_override, Option, #{})
+    ),
+    Channel = #channel{
+        ctx = Ctx,
+        conninfo = ConnInfo,
+        clientinfo = ClientInfo,
+        clientinfo_override = Override,
+        timers = #{},
+        transaction = #{},
+        conn_state = idle,
+        subscriptions = []
+    },
+    Channel1 = init_conn_state(Channel),
+    trigger_post_init(Channel1).
+
+trigger_post_init(Channel) ->
+    case Channel#channel.conninfo of
+        #{conn_mod := emqx_gateway_conn} ->
+            self() ! {timeout, undefined, post_init},
+            Channel;
+        _ ->
+            Channel
+    end.
+
+init_conn_state(Channel = #channel{conninfo = ConnInfo, clientinfo = ClientInfo}) ->
+    case is_auth_required(ClientInfo) of
+        false ->
+            NConnInfo = ConnInfo#{connected_at => erlang:system_time(millisecond)},
+            ClientId = maps:get(clientid, ClientInfo),
+            emqx_logger:set_metadata_clientid(ClientId),
+            Channel#channel{conninfo = NConnInfo, conn_state = anonymous};
+        true ->
+            Channel
+    end.
+
+info_frame(#channel{conninfo = ConnInfo, clientinfo = ClientInfo}) ->
+    {SockHost, SockPort} = maps:get(sockname, ConnInfo),
+    {ok, Vsn} = application:get_key(emqx_gateway_nats, vsn),
+    MsgContent = #{
+        server_id => emqx_conf:get([gateway, nats, server_id]),
+        server_name => emqx_conf:get([gateway, nats, server_name]),
+        version => list_to_binary(Vsn),
+        host => list_to_binary(inet:ntoa(SockHost)),
+        port => SockPort,
+        max_payload => emqx_conf:get([gateway, nats, protocol, max_payload_size]),
+        proto => 0,
+        headers => true,
+        auth_required => is_auth_required(ClientInfo),
+        tls_handshake_first => false,
```

Comment:
should always be `true` when `tls_required` is true, because we do not support data exchange before tls upgrade (unless we inject a upgrade function to esockd)